File: Pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool validity(long long ccn_number);
int check_length(long long ccn_number);
bool check_sum(long long ccn_number);
void print_ccn(long long ccn_number);

int main(void)
{
    long long ccn_number;
    do
    {
        ccn_number = get_long("Number:"); //Request user input credit card number
    }
    while(ccn_number < 0);

    if (validity(ccn_number)) //Print out the type of credit card if number valid
    {
        print_ccn(ccn_number);
    }
    else
    {
        printf("Invalid\n");
    }
}

/* Check validity for credit card number */
bool validity(long long ccn_number)
{
    int ccn_length = check_length(ccn_number);
    if((ccn_length == 13 || ccn_length == 15 || ccn_length == 16) && check_sum(ccn_number))
    {
        return true;
    }
    else
    {
        return false;
    }
}

/* Obtain the number of digits for credit card number */
int check_length(long long ccn_number)
{
    int len;
    for(len = 0; ccn_number > 0 ; len++, ccn_number /= 10);
    return len;
}

/* Perform Luhn's algorithm to check validity */
bool check_sum(long long ccn_number)
{
    int i, sum = 0;
    for(i = 0; ccn_number > 0; i++, ccn_number /= 10)
    {
        if(i % 2 == 0)
        {
            sum += ccn_number % 10;
        }
        else
        {
            int doub;
            doub = 2 * (ccn_number % 10);
            sum += (doub / 10) + (doub % 10);
        }
    }

    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

/* Evaluate type of credit card for input card number and display type */
void print_ccn(long long ccn_number)
{
    if((ccn_number >= 34e13 && ccn_number < 35e13) || (ccn_number >= 37e13 & ccn_number < 38e13))
    {
        printf("AMEX\n");
    }
    else if(ccn_number >= 51e14 && ccn_number < 56e14)
    {
        printf("MASTERCARD\n");
    }
    else if((ccn_number >= 4e12 && ccn_number < 5e12) || (ccn_number >= 4e15 && ccn_number < 5e15))
    {
        printf("VISA\n");
    }
}

File: Pset6/sentimental-credit/credit.py
from cs50 import get_int

# TODO: Request user input credit card number and print out number if valid
def main():

    # Ensure correct usage
    while True:
        ccn_number = get_int("Number: ")
        if ccn_number >= 0:
            break

    if validity(ccn_number):
        print_ccn(ccn_number)
    else:
        print("Invalid")


# Check validity for credit card number
def validity(ccn_number):
    ccn_length = check_length(ccn_number)
    return (ccn_length == 13 or ccn_length == 15 or ccn_length == 16) and check_sum(ccn_number)


# Obtain number of digits for credit card number
def check_length(ccn_number):
    dlen = 0
    for dlen in range(len(str(ccn_number))):
        dlen += 1
    return dlen


# Perform Luhn's algorithm to check validity
def check_sum(ccn_number):
    sum = 0
    # ccn changed to string format to obtain number of digits of card
    # i gets range from 0 to length of str -1
    for i in range(len(str(ccn_number))):
        if (i % 2 == 0):
            sum += ccn_number % 10
        else:
            doub = 2 * (ccn_number % 10)
            # Integer division operator // used to obtain first digit of number eg. 16= 1+6
            sum += (doub // 10) + (doub % 10)
        # Reduce length of ccn_number. Range of i is not recomputed.
        ccn_number //= 10
    return sum % 10 == 0    


# Evaluate type of credit card for input card number and display type
def print_ccn(ccn_number):
    if (ccn_number >= 34e13 and ccn_number < 35e13) or (ccn_number >= 37e13 and ccn_number < 38e13):
        print("AMEX")
    elif ccn_number >= 51e14 and ccn_number < 56e14:
        print("MASTERCARD")
    elif (ccn_number >= 4e12 and ccn_number < 5e12) or (ccn_number >= 4e15 and ccn_number < 5e15):
        print("VISA")
    else:
        print("INVALID")


# Python's way of calling main function after user-defined functions
if __name__ == "__main__":
    main()
